### PR TITLE
Documentation and new algorithm

### PR DESCRIPTION
**Pull request by AUnderwood9**
Added in-line documentation to Program, Grid, Room, Generator, RoomHelper, Binary, and Sidewinder classes. Implemented the Sidewinder algorithm in the Sidewinder class. And added a User input operation to choose between the two algorithm types in the Program class.

### CODE REVIEW

File: CSharp/Mazes/Grid/Grid.cs [hunk 1]
```diff
@@ -2,6 +2,11 @@
 using System.Collections;
 using System.Collections.Generic;
 
+/*
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Grid.cs]: Use /// notation for comments.  

Remove the part about the 2D array.  People can see that I am using a 2D array from the code.

Move the description to `public sealed class Grid`  and not at the namespace level.

File: CSharp/Mazes/Grid/Grid.cs [hunk 2]
```diff
@@ -47,6 +52,10 @@ public int GetColumns()
             return columns;
         }
 
+        /*
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Grid.cs]: Use /// for method comments.

Don't talk about RoomHelper here.  Move the comment to RoomHelper.  Remove the sentence with `starts out as null`.  People can see from the code that roomToConnect with is null and you don't need to explain it again.  

Method level comments should explain what it does, not the how.

File: CSharp/Mazes/Grid/Room.cs [hunk 3]
```diff
@@ -11,12 +11,13 @@ public sealed class Room
     {
         public int row { get; private set; }
         public int column { get; private set; }
-        public int distance { get; set; }
+        public int distance { get; set; }
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Room.cs]: Remove the extra space.

File: CSharp/Mazes/Grid/Room.cs [hunk 4]
```diff
@@ -11,12 +11,13 @@ public sealed class Room
     {
         public int row { get; private set; }
         public int column { get; private set; }
-        public int distance { get; set; }
+        public int distance { get; set; } 
         public bool visited { get; set; }
 
-        private Dictionary<Direction, Room> neighbors;
+        // Keeps track of the neighbors of said room to create a linked list type structure
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Room.cs]: use /// notation to comment properties.  Remove `...to create` and everything after.  Two reasons

1. What happens if I change what datatype I'm using
2. Linked List is not a structure
3. That part does not help explain why the Dictionary is there.

File: CSharp/Mazes/Grid/Room.cs [hunk 5]
```diff
@@ -11,12 +11,13 @@ public sealed class Room
...
+        private Dictionary<Direction, Room> neighbors; 
         
-
+        // Constructor
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Room.cs]: Use /// for method level comments.

File: CSharp/Mazes/Grid/Room.cs [hunk 6]
```diff
@@ -30,12 +31,15 @@ public Room(GridPosition gridPosition)
             this.row = gridPosition.row;
             this.column = gridPosition.column;
 
-            neighbors = new Dictionary<Direction, Room>();
+            // The direction of the room is the key.
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/Room.cs]: This comment repeats what the code says.  Please remove it.

File: CSharp/Mazes/Grid/RoomHelper.cs [hunk 7]
```diff
@@ -9,25 +9,30 @@ public sealed class RoomHelper
...
                 return new GridPosition(room.row + 1, room.column);
             }
-            else if (direction == Direction.EAST)
+            // move back a column
+            else if (direction == Direction.EAST) 
             {
                 return new GridPosition(room.row, room.column + 1);
             }
-            else if (direction == Direction.WEST)
+            // move up a column
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/RoomHelper.cs]: NIT: Can't move up a column.  Move right instead.

File: CSharp/Mazes/Grid/RoomHelper.cs [hunk 8]
```diff
@@ -53,7 +58,8 @@ public static void ConnectRooms(Room roomOne, Room roomTwo, Direction directionF
             }
         }
 
-        public static List<Direction> GetBoundriesRoomIsOn(Room room, int rows, int columns)
+        // This method finds all of the boundaries that each room is on. For example if a room is on the northern wall then it will have a boundary to the north
```
Comment (dhoehna) [inline CSharp/Mazes/Grid/RoomHelper.cs]: Please use /// for method level comments.

File: CSharp/Mazes/MazeAlgorithms/Binary.cs [hunk 9]
```diff
@@ -5,6 +5,10 @@
 using static Grid.Grid;
 using System.Linq;
 
+/*
```
Comment (dhoehna) [inline CSharp/Mazes/MazeAlgorithms/Binary.cs]: Use /// to describe a class.  Please move this to above the class deceleration, not the namespace.

File: CSharp/Mazes/MazeAlgorithms/Sidewinder.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using Grid;
+using System.Collections.Generic;
+using System.Text;
+using static Grid.Grid;
+using System.Linq;
+
+namespace MazeAlgorithms
+{
+    public class Sidewinder : IMazeAlgorithm
```
Comment (dhoehna) [inline CSharp/Mazes/MazeAlgorithms/Sidewinder.cs]: Add a  /// comment here to describe the class.

File: CSharp/Mazes/MazeAlgorithms/Sidewinder.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        private enum ValidConnectionDirections
+        {
+            NORTH = Direction.NORTH,
```
Comment (dhoehna) [inline CSharp/Mazes/MazeAlgorithms/Sidewinder.cs:17]: Good job. :)

File: CSharp/Mazes/MazeGenerator/Generator.cs [hunk 12]
```diff
@@ -17,6 +17,7 @@ public Generator(IGrid gridToManipulate, IMazeAlgorithm algorithmToApply, IDista
             this.solver = solver;
         }
 
+        // The algorithm that is used to create the rooms is applied here.
```
Comment (dhoehna) [inline CSharp/Mazes/MazeGenerator/Generator.cs]: Use /// for comments describing a method.

File: CSharp/Mazes/Mazes/Program.cs [hunk 13]
```diff
@@ -22,31 +22,64 @@ class Program
         const int PEN_THICKNESS_IN_PIXLES = 3;
         static void Main(string[] args)
         {
+
+            // Prompt the user to input the type of algorithm we will be using. 1 for Binary 2 for sidewinder
+            System.Console.WriteLine("Please tell me what type of algorithm you want to use. (1 - Binary 2 - Sidewinder)");
+            int algoType;
```
Comment (dhoehna) [inline CSharp/Mazes/Mazes/Program.cs]: NIT: `algorithmType` not `algoType` The first is more clear.

File: CSharp/Mazes/Mazes/Program.cs [hunk 14]
```diff
@@ -22,31 +22,64 @@ class Program
...
+
+            // Create the base grid
             IGrid grid = new Grid.Grid(ROWS, COLUMNS);
-            IMazeAlgorithm binaryAlgorithm = new Binary();
+            // Determine and create the type of algorithm that will be used to generate the rooms
+            IMazeAlgorithm currentAlgorithm;
```
Comment (dhoehna) [inline CSharp/Mazes/Mazes/Program.cs]: Nice job on finding a way to continually ask for the correct input.

File: CSharp/Mazes/Mazes/Program.cs [hunk 15]
```diff
@@ -22,31 +22,64 @@ class Program
...
+                        currentAlgorithm = new Binary();
+                        break;
+
+                    case 2:
+                        currentAlgorithm = new Sidewinder();
+                        break;
+
+                    default:
+                        System.Console.WriteLine("Can't find this algorithm. Please choose a valid one. (1 - Binary 2 - Sidewinder)");
+                        currentAlgorithm = null;
```
Comment (dhoehna) [inline CSharp/Mazes/Mazes/Program.cs]: NIT: Make currentAlgortihm null when you declare it, so you don't have to re-declare it here.

**Review by dhoehna — CHANGES_REQUESTED** [review]

File: CSharp/Mazes/Grid/Room.cs (hunk 5 above)
Comment (AUnderwood9) [reply CSharp/Mazes/Grid/Room.cs]: Removed this comment. It is not needed.

**Comment by AUnderwood9** [comment]
I have addressed the issues presented here and pushed them to the branch to be merged.

**Review by dhoehna — APPROVED** [review]

**Pull request merged**